Support JSONPath condition without value

minherz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Extends current JSONPath condition logic to support matching "any" value for the provided JSON path expression.
With this change, users will be able to wait on JSON path conditions with values they cannot know ahead including simple or complex (object or array) values. See #117761 for more details.

#### Which issue(s) this PR fixes:

Fixes #117761

#### Special notes for your reviewer:

This PR duplicates in-progress work in #117823. Posted here as an alternative implementation. It will be either merged or closed depending on the collaborative decision with the author of #117823.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The "value" part in the `wait --for=jsonpath='{expression}'[=value]` is now optional. If the value is not provided i.e. the command looks like `wait --for=jsonpath='{expression}'` then the wait condition is interpreted as matched when the expression returns *any* single JSON value like object or a literal.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

minherz committed:
Support JSONPath condition without value

Extend current JSONPath condition logic to return from wait on "any" value.
Change parsing JSONPath input to support the syntax without value.
Match any simple or complex (object or array) values.

minherz commented:
I would like to add a unit testing that validates parsing logic for the flag arguments. I did not find the tests under [wait_test.go]. Is this file the right place for such tests?

[wait_test.go]: https://github.com/kubernetes/kubernetes/blob/1dba84be8a1ced559ff0f02acb876626ebefae93/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go

minherz committed:
chore(fix): test jsonpath condition parsing errors

Test parsing logic for invalid JSONPath condition formats,
excluding JSON path expression parsing.
Fix error in parsing logic

minherz commented:
@ardaguclu can you plz, `/ok-to-test`?

minherz commented:
/retest

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -240,18 +241,23 @@ func newJSONPathParser(jsonPathExpression string) (*jsonpath.JSONPath, error) {
 	}
-	if jsonPathCond == "" {
-		return "", "", errors.New("jsonpath wait condition cannot be empty")
+	if len(jsonPathInput) > 1 {
```
Should this be?;
```suggestion
	if len(jsonPathInput) == 2 {
```

ardaguclu replied:
and I'd prefer early return, such as;
```go
if len(jsonPathInput) == 1 {
    return relaxedJSONPathExp, "", nil
}
```

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -240,18 +241,23 @@ func newJSONPathParser(jsonPathExpression string) (*jsonpath.JSONPath, error) {
-	if jsonPathCond == "" {
-		return "", "", errors.New("jsonpath wait condition cannot be empty")
+	if len(jsonPathInput) > 1 {
+		jsonPathValue := jsonPathInput[1]
```
Could you please add a comment above this to show that an example of what `jsonPathInput[1]` can be?

minherz replied:
I have changed the wording of the error message. Instead of
`jsonpath wait condition cannot be empty`
it says
`jsonpath wait has to have a value after equal sign, like --for=jsonpath='{.status.readyReplicas}'=3`

I am not sure what value the comment with an example of `jsonPathInput[1]` value brings here. Any non-empty string will do (at parsing phase). I looked at "[old code][1]" and it does the same thing: storing a token as a string. It just stores it to a variable with a different name.
Would you agree that both variable names are self-descriptive regarding expectations for the value of `jsonPathInput[1]`?

[1]: https://github.com/kubernetes/kubernetes/blob/8ffbbe455b30d030a8c99857ad85282f6aec95e6/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go#L213

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go:
```diff
@@ -240,18 +241,23 @@ func newJSONPathParser(jsonPathExpression string) (*jsonpath.JSONPath, error) {
+	if len(jsonPathInput) > 1 {
+		jsonPathValue := jsonPathInput[1]
+		if jsonPathValue == "" {
+			return "", "", errors.New("jsonpath wait value cannot be empty")
```
Could you please give an example of how `jsonPathInput[1]` is empty?. Because we checked that length of jsonPathInput is 1 or 2 already.

minherz replied:
I've rewrote the error message to explain how the value can be empty

ardaguclu reviewed (commented):
Thanks for the PR!

Could you please add integration tests in here https://github.com/kubernetes/kubernetes/blob/60401d35d1dcdf4814674ff3fd96b70c312700d0/test/cmd/wait.sh#L21 for the changes?

ardaguclu commented:
/triage accepted
/priority backlog

minherz committed:
chore: address review feedback

add integration test to wait for json without value
refactor JSON condition value parsing and validating
adjusting test to reflect the error message refactoring

minherz commented:
@ardaguclu PTAL

ardaguclu commented on test/cmd/wait.sh:
```diff
@@ -94,6 +94,21 @@ EOF
+    kubectl create deployment test-3 --image=busybox
+
+    # wait with jsonpath without value to succeed
+    set +o errexit
```
I don't think we'd need this, since we are using `2>&1`

minherz replied:
It is not entirely correct. The command `set +o errexit` "tells" the script to not terminate if one of the following executions end with non-zero return code. The `2>&1` merely redirects all prints to `stderr` to be in `stdout`.
If you will look on other lines in this script or into other scripts, you will see that they all follow similar pattern.

ardaguclu replied:
You are right. Especially considering that the few lines above was added by me with the same :)

ardaguclu commented on test/cmd/wait.sh:
```diff
@@ -94,6 +94,21 @@ EOF
+    # wait with jsonpath without value to succeed
+    set +o errexit
+    # Command: Wait with jsonpath without value
+    output_message=$(kubectl wait --for=jsonpath='{.status.replicas}' deploy/test-3 2>&1)
```
What would happen, if I run this test on master branch(without your patch)?
What would happen, if I run this command without `2>&1`?

minherz replied:
> What would happen, if I run this test on master branch(without your patch)?

The test will fail because the matching output will be different. It would say that the format is wrong and the JSON condition is missing a value.

> What would happen, if I run this command without `2>&1`?

I am not entirely sure because I do not know what output is redirected to `stderr`.

ardaguclu reviewed (commented):
Changes looks good to me, just a few questions

minherz commented:
@ardaguclu I've resolved previous comments and addressed the questions from your last feedback. PTAL

ardaguclu commented:
/approve
/lgtm
